src/charm.py

```diff
@@ -176,9 +137,26 @@ def _check_model_name(self):
 
     def _check_leader(self):
         if not self.unit.is_leader():
-            # We can't do anything useful when not the leader, so do nothing.
             raise CheckFailed("Waiting for leadership", WaitingStatus)
 
+    def _update_layer(self, profiles_service: str) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        current_layer = self._container.get_plan()
+        new_layer = self.get_kubeflow_dashboard_operator_layer(profiles_service)
+        self.logger.info(f"NEW LAYER: {new_layer}")
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info(
+                    "Pebble plan updated with new configuration, restarting"
+                )
+                self._container.restart(self._container_name)
+            except ChangeError as e:
+                self.logger.error(traceback.format_exc())
+                self.unit.status = BlockedStatus("Failed to restart")
+                raise e
```

Comment:
Have you encountered this case in a full deployment and it done what you expect?  If not, test it see what gets printed to logs and what status is set.  I might be wrong on what will happen, but I think you want either the first two lines or the third, but not both.

What I expect is:
* `self.logger.error()` will print the traceback to juju debug-log
* `self.unit.status = BlockedStatus()` will set the status
* `raise e` (if uncaught by anything above, which I don't see) will cause Juju to print the error message to debug-log (as second time) and set status to `ErrorStatus` (covering over the `BlockedStatus` and requiring the user to intervene with a `juju resolved`)

Reply:
If you want to raise `BlockedStatus` for this situation, think about using the same pattern as `_check_container_connection` and others